ext/storage_sources/s3_store/aws_bucket_conn.cpp

```diff
@@ -0,0 +1,123 @@
+#include <aws/core/Aws.h>
+#include <aws/s3-crt/S3CrtClient.h>
+#include <iostream>
+#include <fstream>
+
+#include "aws_bucket_conn.h"
+#include <aws/s3-crt/model/ListObjectsRequest.h>
+#include <aws/s3-crt/model/PutObjectRequest.h>
+#include <aws/s3-crt/model/DeleteObjectRequest.h>
+
+#include <aws/s3/model/Object.h>
+#include <vector>
+#include <string>
+#include <sys/stat.h>
+
+bool
```

Comment:
I would think we sould return a `vector`, otherwise, we may want to call `clear` on the `vector` passed to the function but clearing a vector defined by a user does not seem good practice to me.

Reply:
I agree we should return a bucket. This looks like c style code not c++ style.

Reply:
@quchenhao , should we throw an exception or die if we get to the `return false` section?

Reply:
I think we should return exception

Reply:
Thanks @quchenhao.
The function should:

- Return a `std::vector`
- Not take any argument
- An exception should be thrown if `outcome.IsSuccess()` is evaluated to false. We cold start with `throw "<message>"`

@sulabhM , does it sound ok?

Reply:
List buckets will actually go away, it is temporary just to allow us to test and build other infrastructure pieces. Hence, I am avoiding fixing the style and handling errors for now. We will still need to have these discussions for other methods that will form the part of the final code, eg: list-objects.

But I am glad you are already participating, we should have a discussion on what these functions will look like in the upcoming tickets. Overall it sounds okay to me to throw an exception in case of an error, with the message. I am not sure of returning a vector, I would want to avoid a copy when returning. I will have to read more about the optimizations for return and move operator otherwise - in the upcoming tickets.

Reply:
return should trigger the move constructor of vector, which should avoid doing a full copy.

Reply:
As discussed, we can leave that function as is. It is sufficient to keep progressing. Any rework will be done at a later stage to suit the new needs/requirements.